### PR TITLE
add facilities check on mobile appts

### PR DESCRIPTION
**Pull request by aherzberg**
## Summary
Add Facilities check to VAOS policy. This will not effect VAOS controller.

created separate ticket to merge VAOS policy logic into same place: https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/9617

## Related issue(s)
https://github.com/department-of-veterans-affairs/va-mobile-app/issues/9415

## Testing done

- [ ] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: modules/vaos/app/policies/vaos_policy.rb
Comment (aherzberg) [inline modules/vaos/app/policies/vaos_policy.rb:1]: This is a separate check so we can return a unique error that FE can eventually use to trigger a banner saying that they don't have any facilities.

File: modules/mobile/spec/requests/mobile/v0/appointments/create_spec.rb [hunk 1]
```diff
@@ -16,6 +16,7 @@
   let!(:user) { sis_user(icn: '1012846043V576341') }
 
   before do
+    allow_any_instance_of(User).to receive(:va_patient?).and_return(true)
```
Comment (aherzberg) [inline modules/mobile/spec/requests/mobile/v0/appointments/create_spec.rb:19]: just adding the facilities was not working in this spec as it was transforming facilities into test facilities 984. This change is unique to testing env so not worried about effecting prod performance.

File: modules/mobile/app/controllers/mobile/v0/appointments_controller.rb [hunk 2]
```diff
@@ -137,6 +137,7 @@ def appointments_cache_interface
       def authorize
         raise_access_denied unless current_user.authorize(:vaos, :access?)
         raise_access_denied_no_icn if current_user.icn.blank?
+        raise_access_denied_no_facilities unless current_user.authorize(:vaos, :facilities?)
```
Comment (aherzberg) [inline modules/mobile/app/controllers/mobile/v0/appointments_controller.rb:140]: @kpethtel , I'm unsure if this warrants a flipper. It seems simple enough but let me know if you feel differently.

File: modules/vaos/app/policies/vaos_policy.rb [hunk 3]
```diff
@@ -4,4 +4,8 @@
   def access?
     Flipper.enabled?('va_online_scheduling', user) && user.loa3?
   end
+
+  def facilities?
```
Comment (kpethtel) [inline modules/vaos/app/policies/vaos_policy.rb:8]: Would `facilities_access?` make sense here?

**Review by kpethtel — APPROVED** [review]
Looks good. 

I don't think this merits a feature flag. We've been told this is the right thing to do and it makes conceptual sense. If it causes issues, those issues won't be 500s. They'll be user complaints. If we receive enough of those, we can stop checking the new policy and keep asking questions about why this isn't the right thing to do. 

Is vaos also going to use the new policy?

File: modules/vaos/app/policies/vaos_policy.rb (hunk 3 above)
Comment (aherzberg) [reply modules/vaos/app/policies/vaos_policy.rb]: yeah thats better. Updated.

**Comment by aherzberg** [comment]
> Looks good.
> 
> I don't think this merits a feature flag. We've been told this is the right thing to do and it makes conceptual sense. If it causes issues, those issues won't be 500s. They'll be user complaints. If we receive enough of those, we can stop checking the new policy and keep asking questions about why this isn't the right thing to do.
> 
> Is vaos also going to use the new policy?

I created a new ticket to have VAOS use as our policy logic is near identical so I was thinking of merging it into a single place but it gets a little more complicated since it's part of their parent controller and we can't just inherit that controller since we have different grandparent controllers. I could do it within this PR but figured if we did, it'd be a lot more changes with the VAOS spec changes. 

https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/9617

@simiadebowale How do you feel about moving the VAOS base controller to a concern that Mobile and VAOS can share? VAOS can also use this facilities check even though it was be redundant with the check being done online. Hopefully we can move to the logic only being on the BE.

**Review by kpethtel — DISMISSED** [review]
A controller concern would be cleaner and the VAOS team can add that if they want (since it should be in their module), but I don't think it's a big deal. Looks good.

**Review by kpethtel — APPROVED** [review]

**Comment by ryan-mcneil** [comment]
@aherzberg please pull master to take care of the CVE

**Review by ryan-mcneil — DISMISSED** [review]

**Comment by ryan-mcneil** [comment]
@aherzberg let me know when you're ready to merge - I might have to override because of that EOF error

**Comment by aherzberg** [comment]
@ryan-mcneil should be good to merge now. Looks like prod deploy has already gone out for the day so I should have the rest of the day to test it. Thanks!

**Comment by ryan-mcneil** [comment]
@aherzberg Ope maybe I missed it before, but there are also failing specs... are any of those potentially related to your changes? Otherwise let's take note of which ones are failing so I can talk to the respective teams and get eyes on them. Thanks!

**Comment by aherzberg** [comment]
@ryan-mcneil I wasn't able to reproduce the errors locally on all 3 failures so was assuming it was all environment issues. I've been seeing a lot of that in the last week or so.

**Review by ryan-mcneil — APPROVED** [review]

**Pull request merged**